refactor(linter): consolidate file loading logic

DonIsaac opened the pull request:
# Human Description
Low on time, so this one is short.
- consolidate source file and partial loader logic into `loader` module. I have more plans for this.
- ~LSP no longer uses `VALID_EXTENSIONS`, so now `.d.ts` files (and the like) will be linted as well~ LSP does not respect `.gitignore` files, so this change was reverted.

# AI Description
## Refactor Loader and Partial Loader

This PR refactors the loader and partial loader functionality in the oxc_linter crate:

* Introduce a new `Loader` struct with methods for checking if a file can be loaded and loading file contents
* Move `partial_loader` module to `loader/partial_loader`
* Rename `JavaScriptSource` to `source.rs` and move it to the `loader` module
* Update `JavaScriptSource` to use `u32` for `start` offset instead of `usize`
* Refactor `IsolatedLintHandler` to use the new `Loader`
* Update imports and module references throughout the codebase

This change improves the organization of the loader-related code and provides a more unified interface for loading different file types.

DonIsaac commented:
* **#6130** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6130?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DonIsaac commented:
@Boshen I want to use `oxc_resolver`'s FS abstraction so that we can plug the linter right into VSCode's file API. This will also help with integrating with Monaco. This cant be done unless `FileSystemOs` is exposed publically.

Boshen reviewed (commented):
> so now .d.ts files (and the like) will be linted as well

We should never lint `.d.ts` files from my experience, they are never hand written.

qweered commented:
I heard that some companies write them by hand

DonIsaac commented:
@Boshen this change (linting `.d.ts`) files only affects the LSP, not oxlint. If a user is opening a file in their editor, and it's not already ignored by version control, it's more than likely to be a source file.

Now I need to go back and check that `oxc_language_server` respects .gitignores...

DonIsaac commented:
Update: The LSP does _not_ respect gitignores, so I've re-added the `VALID_EXTENSIONS` check. Note that it looks like the LSP also doesn't support .eslintignore files. We should add this in the future.

DonIsaac committed:
refactor(linter): consolidate file loading logic (#6130)

# Human Description
Low on time, so this one is short.
- consolidate source file and partial loader logic into `loader` module. I have more plans for this.
- ~LSP no longer uses `VALID_EXTENSIONS`, so now `.d.ts` files (and the like) will be linted as well~ LSP does not respect `.gitignore` files, so this change was reverted.

# AI Description
## Refactor Loader and Partial Loader

This PR refactors the loader and partial loader functionality in the oxc_linter crate:

* Introduce a new `Loader` struct with methods for checking if a file can be loaded and loading file contents
* Move `partial_loader` module to `loader/partial_loader`
* Rename `JavaScriptSource` to `source.rs` and move it to the `loader` module
* Update `JavaScriptSource` to use `u32` for `start` offset instead of `usize`
* Refactor `IsolatedLintHandler` to use the new `Loader`
* Update imports and module references throughout the codebase

This change improves the organization of the loader-related code and provides a more unified interface for loading different file types.
